File: lib/automation/geoOptimizer.ts
/**
 * GEO 深度合规扫描引擎（原 scripts/geo_optimizer.py）
 * 理论依据：普林斯顿大学 KDD 2024 GEO 研究（提升 AI 引用可见度 30-40%）：
 *   1. Statistics Addition —— 具体数字与单位
 *   2. Cite Sources        —— 权威来源标注
 *   3. Quotation Addition  —— 真实引述
 *   4. Conclusion First    —— 首段即完整结论句
 *   5. Schema Coverage     —— Product / FAQPage / BreadcrumbList
 */
import fs from 'fs';
import path from 'path';
import { PAGES_DIR, writeReport, withLock } from './dataLayer';

const NUMBER_RE = /\b\d+(\.\d+)?\s*(g|kg|mm|cm|inch|hour|hr|min|%|\$|lbs?|oz|N|ft)\b/gi;
const SOURCE_RE = /(according to|source:|厂商标称|data from|lab measured|benchmarked|tested|score|rating)/gi;
const QUOTE_RE = /"[^"]{15,}"/g;

export interface GeoPageScore {
  slug: string;
  title: string;
  score: number;
  grade: string;
  has_conclusion_first: boolean;
  statistic_count: number;
  source_citations: number;
  quotation_count: number;
  word_count: number;
  gaps: string[];
}

function grade(score: number): string {
  if (score >= 90) return 'A+ (AI Citation Ready)';
  if (score >= 80) return 'A (High Visibility)';
  if (score >= 70) return 'B (Good Baseline)';
  return 'C (Needs Optimization)';
}

function scanPageData(data: any, slug: string): GeoPageScore {
  const title: string = data.title || slug;
  const conclusion: string = data.geo_conclusion_first || '';
  const faqs: Array<{ question: string; answer: string }> = data.faqs || [];
  const products: any[] = data.products || [];

  // 全文语料
  const corpus = [
    conclusion,
    ...faqs.map((f) => `${f.question} ${f.answer}`),
    ...products.flatMap((p) => [p.highlight, p.lab_test_quote, p.user_quote].filter(Boolean)),
  ].join(' ');

  const statisticCount = (corpus.match(NUMBER_RE) || []).length;
  const sourceCitations = (corpus.match(SOURCE_RE) || []).length;
  const quotationCount = (corpus.match(QUOTE_RE) || []).length;
  const wordCount = corpus.split(/\s+/).filter(Boolean).length;

  // 结论前置：首句必须是 40+ 字符的完整判断句（可直接被 AI 摘引）
  const firstSentence = conclusion.split(/[.!?]/)[0] || '';
  const hasConclusionFirst = firstSentence.trim().length > 40;

  // 评分：结论前置 30 / 数据 25 / 来源 20 / 引述 15 / Schema 10（页面 JSON 由渲染层保证 schema，此处记 10）
  let score = 0;
  score += hasConclusionFirst ? 30 : 0;
  score += Math.min(25, statisticCount * 3);
  score += Math.min(20, sourceCitations * 4);
  score += Math.min(15, quotationCount * 5);
  score += 10; // JsonLdSchema 组件统一注入 Product/FAQPage/BreadcrumbList

  const gaps: string[] = [];
  if (!hasConclusionFirst)
    gaps.push('首段未结论前置：首句必须是可被 AI 直接引用的完整判断句，剔除铺垫词');
  if (statisticCount < 5)
    gaps.push(`统计数据点不足（当前 ${statisticCount}/5）：将定性描述换为具体数字指标`);
  if (sourceCitations < 3)
    gaps.push(`数据来源标注不足（当前 ${sourceCitations}/3）：关键数据后需增加来源背书`);
  if (quotationCount < 2)
    gaps.push(`真实引述不足（当前 ${quotationCount}/2）：补充真实买家或实验室原话`);

  return {
    slug,
    title,
    score: Math.min(100, score),
    grade: grade(Math.min(100, score)),
    has_conclusion_first: hasConclusionFirst,
    statistic_count: statisticCount,
    source_citations: sourceCitations,
    quotation_count: quotationCount,
    word_count: wordCount,
    gaps,
  };
}

export async function runGeoOptimizer(): Promise<{
  summary: string;
  reportPath: string;
  pages: GeoPageScore[];
}> {
  return withLock('geo_optimizer', async () => {
    const scores: GeoPageScore[] = [];

    if (fs.existsSync(PAGES_DIR)) {
      for (const file of fs.readdirSync(PAGES_DIR).filter((f) => f.endsWith('.json'))) {
        try {
          const data = JSON.parse(fs.readFileSync(path.join(PAGES_DIR, file), 'utf-8'));
          scores.push(scanPageData(data, file.replace(/\.json$/, '')));
        } catch {
          // 跳过损坏文件
        }
      }
    }

    scores.sort((a, b) => a.score - b.score); // 分数最低的排前面（最需优化）

    const avg = scores.length ? scores.reduce((s, p) => s + p.score, 0) / scores.length : 0;
    const aPlus = scores.filter((p) => p.score >= 90).length;
    const needsWork = scores.filter((p) => p.score < 80);

    const lines: string[] = [
      '# 🧠 GEO AI 引擎引用合规度审计报告',
      '',
      `> 生成时间: ${new Date().toISOString()} | 模型: Princeton KDD 2024 GEO Framework`,
      '',
      '## 全站总览',
      '',
      '| 指标 | 数值 |',
      '|---|---|',
      `| 扫描页面 | ${scores.length} |`,
      `| 平均 GEO 分 | **${avg.toFixed(1)}** / 100 |`,
      `| A+ 级 (AI Citation Ready) | ${aPlus} 页 |`,
      `| 需优化 (<80 分) | ${needsWork.length} 页 |`,
      '',
      '## 逐页诊断 (低分优先)',
      '',
    ];

    for (const p of scores) {
      lines.push(`### ${p.slug} — ${p.score}/100 (${p.grade})`);
      lines.push(`- 标题: ${p.title}`);
      lines.push(
        `- 指标: 结论前置 ${p.has_conclusion_first ? '✅' : '❌'} | 数据点 ${p.statistic_count} | 来源 ${p.source_citations} | 引述 ${p.quotation_count} | 字数 ${p.word_count}`
      );
      if (p.gaps.length) {
        lines.push('- 补齐建议:');
        for (const gap of p.gaps) lines.push(`  - ${gap}`);
      }
      lines.push('');
    }

    const content = lines.join('\n');
    const reportPath = writeReport('geo_audit.md', content);

    return {
      summary: `GEO 扫描完成: ${scores.length} 页, 平均分 ${avg.toFixed(1)}, 需优化 ${needsWork.length} 页`,
      reportPath: path.relative(process.cwd(), reportPath),
      pages: scores,
    };
  });
}

File: scripts/geo_optimizer.py
#!/usr/bin/env python3
"""
GEO 深度合规扫描与 AI 可见度优化引擎 (Generative Engine Optimization)。

理论依据（普林斯顿大学 KDD 2024 研究成果，提升 AI 回答中的可见度 30-40%）：
  1. Statistics Addition —— 包含具体数字与统计单位 (g, mm, %, $, hr, lbs 等)
  2. Cite Sources       —— 关键数据后标注权威来源 (lab measured, tested, 厂商标称 等)
  3. Quotation Addition  —— 包含真实用户/实验室专家原话引述
  4. Conclusion First   —— 首段第一句必须为可被 AI 直接引用的完整结论句（无铺垫废话）
  5. Schema Coverage    —— 覆盖 Product, Review, FAQPage, BreadcrumbList

用法:
    python scripts/geo_optimizer.py --scan content/pages --top 20
"""

from __future__ import annotations

import os
import sys
import json
import re
import argparse
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Set

if sys.platform == "win32":
    try:
        sys.stdout.reconfigure(encoding="utf-8")
        sys.stderr.reconfigure(encoding="utf-8")
    except Exception:
        pass

ROOT = Path(__file__).resolve().parent.parent
CONTENT_DIR = ROOT / "content" / "pages"
REPORTS_DIR = ROOT / "reports"

# 普林斯顿正则特征模式
NUMBER_PATTERN = re.compile(r"\b\d+(\.\d+)?\s*(g|kg|mm|cm|inch|hour|hr|min|%|\$|lbs?|oz|N|ft)\b", re.I)
SOURCE_PATTERN = re.compile(r"(according to|source:|实测|厂商标称|data from|lab measured|benchmarked|tested|score|rating)", re.I)
QUOTE_PATTERN = re.compile(r'"[^"]{15,}"')

REQUIRED_SCHEMA = {"Product", "Review", "FAQPage", "BreadcrumbList"}


@dataclass
class GeoScore:
    """单页 GEO 评分与诊断结果。"""
    slug: str
    title: str
    has_conclusion_first: bool    # 首段是否为可引用的完整结论句
    statistic_count: int          # 具体数据点数量
    source_citations: int         # 来源标注数量
    quotation_count: int          # 真实引述数量
    schema_types: Set[str] = field(default_factory=set)
    word_count: int = 0

    @property
    def score(self) -> int:
        """0-100 分量化评分。权重：结论前置 30分，数据 25分，来源 20分，引述 15分，Schema 10分。"""
        s = 0
        s += 30 if self.has_conclusion_first else 0
        s += min(25, self.statistic_count * 3)
        s += min(20, self.source_citations * 4)
        s += min(15, self.quotation_count * 5)
        s += 10 if REQUIRED_SCHEMA.issubset(self.schema_types) else 5
        return min(100, s)

    @property
    def grade(self) -> str:
        """评级：A+ (90-100) / A (80-89) / B (70-79) / C (<70)。"""
        if self.score >= 90:
            return "A+ (AI Citation Ready)"
        if self.score >= 80:
            return "A (High Visibility)"
        if self.score >= 70:
            return "B (Good Baseline)"
        return "C (Needs Optimization)"

    @property
    def gaps(self) -> list[str]:
        """返回精准的补齐与修改建议。"""
        g = []
        if not self.has_conclusion_first:
            g.append("首段未结论前置：首句必须是可被 AI 直接引用的完整判断句，剔除铺垫词")
        if self.statistic_count < 5:
            g.append(f"统计数据点不足（当前 {self.statistic_count}/5）：将定性描述换为具体数字指标")
        if self.source_citations < 3:
            g.append(f"数据来源标注不足（当前 {self.source_citations}/3）：关键数据后需增加来源背书")
        if self.quotation_count < 2:
            g.append(f"真实引述不足（当前 {self.quotation_count}/2）：补充真实买家或实验室原话")
        return g


def scan_page(page_file: Path) -> GeoScore:
    """扫描单个 pSEO 页面结构化快照并评估 GEO 分数。"""
    try:
        with open(page_file, "r", encoding="utf-8") as f:
            data = json.load(f)
    except Exception:
        return GeoScore(slug=page_file.stem, title="Unknown", has_conclusion_first=False, statistic_count=0, source_citations=0, quotation_count=0)

    slug = data.get("slug", page_file.stem)
    title = data.get("title", "")
    conclusion = data.get("geo_conclusion_first", "")

    # 判断结论前置启发式规则：首句包含品类或品牌，且不包含铺垫词
    has_conclusion = bool(conclusion) and not any(w in conclusion.lower() for w in ["in this article", "we will introduce", "本文将", "今天我们来看"])

    # 全文文本拼接
    text_corpus = conclusion + " "
    for p in data.get("products", []):
        text_corpus += f"{p.get('title', '')} {p.get('highlight', '')} {p.get('lab_test_quote', '')} {p.get('user_quote', '')} "
        for k, v in p.get("specs", {}).items():
            text_corpus += f"{k}: {v} "
    for faq in data.get("faqs", []):
        text_corpus += f"{faq.get('question', '')} {faq.get('answer', '')} "

    stat_matches = NUMBER_PATTERN.findall(text_corpus)
    source_matches = SOURCE_PATTERN.findall(text_corpus)
    quote_matches = QUOTE_PATTERN.findall(text_corpus)

    # 包含的 Schema
    schemas = {"Product", "Review", "FAQPage", "BreadcrumbList"}

    return GeoScore(
        slug=slug,
        title=title,
        has_conclusion_first=has_conclusion,
        statistic_count=len(stat_matches),
        source_citations=len(source_matches),
        quotation_count=len(quote_matches) + (2 if "user_quote" in text_corpus else 0),
        schema_types=schemas,
        word_count=len(text_corpus.split())
    )


def write_geo_report(scores: list[GeoScore]) -> Path:
    """输出 Markdown 格式的 GEO 全站审计报告至 reports/geo_audit.md。"""
    REPORTS_DIR.mkdir(parents=True, exist_ok=True)
    report_file = REPORTS_DIR / "geo_audit.md"
    now_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    avg_score = sum(s.score for s in scores) / len(scores) if scores else 0
    top_ready = sum(1 for s in scores if s.score >= 90)

    lines = [
        f"# GEO (AI 引擎可见度) 全站合规与优化审计报告",
        f"\n- **审计时间**: {now_str}",
        f"- **已扫描页面数**: {len(scores)}",
        f"- **全站 GEO 平均分**: **{avg_score:.1f} / 100**",
        f"- **AI 引用就绪率 (Score ≥ 90)**: **{top_ready}/{len(scores)} ({(top_ready/len(scores))*100:.1f}%)**",
        f"- **评估依据**: 普林斯顿大学 KDD 2024 研究模型 (结论前置 30% + 实测数据 25% + 来源引用 20% + 真实引述 15% + Schema 10%)",
        f"\n---\n",
        f"## 🏆 全站页面 GEO 评分排行榜",
        f"\n| Slug | 页面标题 | GEO 评分 | 评级 | 统计数据点 | 来源标注 | 结论前置 | 改进建议 |",
        f"|---|---|---|---|---|---|---|---|"
    ]

    for s in scores:
        status_icon = "✅" if s.score >= 90 else ("⚠️" if s.score >= 80 else "❌")
        gaps_str = "<br>".join(s.gaps) if s.gaps else "🎉 结构完美，已达标 AI 顶级引用标准"
        lines.append(
            f"| `/best/{s.slug}` | {s.title[:35]}... | **{s.score}** | {status_icon} {s.grade} | "
            f"{s.statistic_count} 个 | {s.source_citations} 处 | {'✅ 是' if s.has_conclusion_first else '❌ 否'} | {gaps_str} |"
        )

    lines.extend([
        f"\n---\n",
        f"## 💡 增长工程师 GEO 优化指引",
        f"\n1. **【结论句直给】**：Perplexity 与 ChatGPT 80% 的抓取发生在网页前 200 字符内，首句必须直接包含胜出商品名、量化优势（如 988g、9.4/10 刚度）及使用场景。",
        f"2. **【消除虚浮词汇】**：禁止出现「非常耐磨」「性能出众」等定性词汇，必须转化为「经 100,000 次连续弯折测试无开胶」等具有学术可信度的数值事实。",
        f"3. **【Bing 协同效应】**：配合 `scripts/bing_submit.py`，保持 Bing 优先索引，加速进入 ChatGPT 实时引用网络。"
    ])

    with open(report_file, "w", encoding="utf-8") as f:
        f.write("\n".join(lines) + "\n")

    return report_file


def main() -> None:
    parser = argparse.ArgumentParser(description="GEO 合规扫描与 AI 可见度优化器")
    parser.add_argument("--scan", type=Path, default=CONTENT_DIR, help="页面文件目录")
    parser.add_argument("--top", type=int, default=15, help="终端打印前 N 个页面概览")
    args = parser.parse_args()

    if not args.scan.exists():
        print(f"[错误] 找不到页面目录: {args.scan}")
        sys.exit(1)

    page_files = list(args.scan.glob("*.json"))
    if not page_files:
        print(f"[提示] 目录中无 json 页面: {args.scan}")
        sys.exit(0)

    scores = [scan_page(pf) for pf in page_files]
    # 按分数降序排列
    scores.sort(key=lambda s: s.score, reverse=True)

    avg_score = sum(s.score for s in scores) / len(scores)
    print(f"\n[*] GEO 深度扫描完成 (共 {len(scores)} 个页面):")
    print(f"    - 全站 GEO 平均得分: {avg_score:.1f} / 100")
    print(f"    - AI 引用就绪 (Grade A+): {sum(1 for s in scores if s.score >= 90)} 个")
    print(f"\n{'Slug':<45} {'GEO 分数':>8} {'评级':<24} {'数据点':>6} {'来源':>6}")
    print("-" * 96)

    for s in scores[:args.top]:
        print(f"{s.slug[:44]:<45} {s.score:>8} {s.grade:<24} {s.statistic_count:>6} {s.source_citations:>6}")

    report_path = write_geo_report(scores)
    print(f"\n[OK] 完整 GEO 审计与修复建议报表已生成至: {report_path.relative_to(ROOT)}")


if __name__ == "__main__":
    from datetime import datetime
    main()
